### DISCUSSION TITLE
Welcome to WiringPi Discussions!

### DISCUSSION THREAD

**Discussion by mwallner** [WiringPi/WiringPi#197] [Announcements]
👋 Welcome!

We’re using Discussions as a place to connect with other members of our community. We hope that you:

* Ask questions you’re wondering about.
* Share ideas.
* Engage with other community members.
* Welcome others and are open-minded. Remember that this is a community we build together 💪.

**Comment by pedross3** [comment]
Hello, world!

I’m trying to communicate between my RPi 4 and my Battery Monitor system (BQ769142) but with no success.
My logic analyzer says that there’s no ack bit.
 
The communication works as follows:
•	First you write the subcommand (in this case, 0x0071) in little endian format to 0x3E, and then 0x3F. Which means, you first write 0x71 to 3E, then 00 to 3F.
•	Then you read the message from the buffer 0x40 and 0x41.
 
My RPi can identify the BM with i2cdetect, and I’m sure the address (which is 0x08) is correct.
I also made sure to send the correct message, with the correct size in bits.
The pull-up resistor required for the i2c is also present.
I’ve already reinstalled the library, as well as check voltage levels to make sure they are compatible.
Clock speed is also correct (400k hz)
 
Please tell me what I’m missing.
 
Thank you in advance,

```
//Here's the code:

#include <iostream>
#include <wiringPiI2C.h>
#include <wiringPi.h>
#include <unistd.h>
#include <stdlib.h>
#include <cstdint>
#include <ctime>

int fd;

int main()
{
	int reg_holder;
	unsigned char subcommand = 0x0071;


	// Initialize wiringPi and I2C
	int fd = wiringPiI2cSetup(0x08);
	if (fd == -1)
	{
		std::cout << "Falsch kommunikation";
		return -1;
	}

	std::cout << "Init result: " << fd << std:endl; // this prints 3 in the cmd

	int reg_addr;

	std::cout << "int main(): i2c setup successfull" << std::endl;

	int i = 0;

	while (1)
	{
		i++;
		reg_addr = 0x3E;
		wiringPiI2CWriteReg8(fd, (reg_addr), subcommand & 0xFF);
		delay(20);



		reg_addr = 0x3F;
		wiringPiI2CWriteReg8(fd, (reg_addr), (subcommand>>8) & 0xFF);
		delay(20);

		reg_holder = 0xFF & 0x40;
		std::cout << "First Address: " << std::hex << reg_holder << std::endl;

		int test1 = wiringPiI2CReadReg16(fd, reg_holder);
		delay(100);

		reg_holder = 0xFF & 0x41;
		int test2 = wiringPiI2cReadReg16(fd, reg_holder);
		delay(10000); // 10 seconds to slow down data transfer when debugging
	}
}
```
![image](https://github.com/WiringPi/WiringPi/assets/26482622/ea3f7b9a-34ba-496c-a4de-62ce6faccd67)

**Reply by mwallner** [reply]
Hi @pedross3 :wave: ,
While I cannot directly assist you with this issue, I think you're chances of getting support are better at 
https://forum-raspberrypi.de/forum/ (German, guess that works for you based on some of the comments in your sourcecode?) or https://forums.raspberrypi.com/ (English).

That said, if anyone stumbles upon this who actually *can* help out - please do!

**Comment by seamusdemora** [comment]
Adding Discussions is a good idea IMO. I do have three questions: 

1. does wiringPi currently support `bookworm` on RPi5? 
2. is the wiringPi team using - or planning on using - the libgpiod library? 
3. what are the plans for continuing wiringPi's `gpio` (command line/scriptable) tools?

Would also like to add my Thanks & Appreciation to all of you for resurrecting this great old library!

**Reply by mstroh76** [reply]
1. yes, Pi 5 reading , writing , paddrive, ISR is ok, PWM feature is missing ; I think it's only accessible as root :-( 

Next thing is checking with Ubuntu LTS.

2. as long as sysfs is working it's not needed. 
Maybe we should replace the sysfs in the future with libgpiod functions. 
But in reality I can't see any 
advantage to DMA. 
What do you think about it?

3. no major changes. Maybe I will include more RP1 functionality. The new chip is quite different. 
Why do you ask? Any suggestions?
The pinctrl is not looking good to me, people like the readall !

**Comment by uugear** [comment]
It is so nice to have WringPi back, and thank you for the great work!

Our software uses the "gpio" utility, we only use it for rather simple and slow scenarios like read, write and wfi. During the days that we don't have a working WringPi on Raspberry Pi, we made a [wrapper for raspi-gpio](https://github.com/uugear/gpio-util) to emulate the output from the gpio utilitiy. It works, but the performance is poor and we have to adapt the changes made in raspi-gpio. Also raspi-gpio will not support Raspberry Pi 5, so we have to find a way.

We previously thought of migrating to pinctrl. Yesterday I stumbled across this repository and I think moving back to wiringPi may be a much better choice :D

At the moment I have two questions: will we be able to install wiringPi via apt or apt-get soon? Is it an allowed/recommended way for us to compile the "gpio" utility and include it into our software package (will credit the work from you in documentation)?

**Reply by mstroh76** [reply]
1.) We want to bring 3.x to Ubuntu and Raspberry Pi OS, but it's not sure whether it will succeed. 
2.) see LGPL, include link to  https://github.com/WiringPi/WiringPi/

**Comment by MaciejSzpakowski** [comment]
I'm trying to compile my "hello world" wiringPi program:

```
#include <stdio.h>
#include <wiringPi.h>

int main()
{

  if(wiringPiSetup() == -1)
  {
    printf("wiringpi fail\n");
    return 0;
  }

  printf("Hello pie\n");
  return 0;
}
```


But I get linking errors:
```
/usr/bin/ld: /usr/lib/gcc/arm-linux-gnueabihf/10/../../../libwiringPi.so: undefined reference to `pthread_join@GLIBC_2.34'
/usr/bin/ld: /usr/lib/gcc/arm-linux-gnueabihf/10/../../../libwiringPi.so: undefined reference to `stat@GLIBC_2.33'
/usr/bin/ld: /usr/lib/gcc/arm-linux-gnueabihf/10/../../../libwiringPi.so: undefined reference to `shm_open@GLIBC_2.34'
/usr/bin/ld: /usr/lib/gcc/arm-linux-gnueabihf/10/../../../libwiringPi.so: undefined reference to `pthread_cancel@GLIBC_2.34'
/usr/bin/ld: /usr/lib/gcc/arm-linux-gnueabihf/10/../../../libwiringPi.so: undefined reference to `pthread_create@GLIBC_2.34'
```
This is how I'm trying to compile:
```
gcc main.c -lwiringPi -lpthread
```
Interestingly enough, if I use anything from pthread.h by itself (without wiringPi) it compiles fine. What's happening ?

**Comment by mstroh76** [comment]
Please post output:
```
lsb_release -a
uname -a
gpio -v
```
How did you install WiringPi?

**Comment by seamusdemora** [comment]
@mstroh76 :
> 
>     2. as long as sysfs is working it's not needed.
>        Maybe we should replace the sysfs in the future with libgpiod functions.
>        But in reality I can't see any
>        advantage to DMA.
>        What do you think about it? 

What everyone has been *saying* about `sysfs` is that it just won't work after the 6.6 Linux kernel is adopted. I don't know enough about that to offer an opinion, but there does seem to be a pretty strong consensus on this point. Do you guys have a plan for this?

> 
>     3. no major changes. Maybe I will include more RP1 functionality. The new chip is quite different.
>        Why do you ask? Any suggestions?
>        The pinctrl is not looking good to me, people like the readall !

I was just curious about `gpio` because it was the only part of wiringPi that I used frequently. The only suggestion that springs to mind is to *consider* revising the pin numbering system. 

Not sure I grok your comment re `pinctrl`. I used it only briefly & found it quite decent - certainly much better than `libgpiod`.
But - what is "readall"??

**Comment by mstroh76** [comment]
Maybe next kernel 6.6. still has sysfs: https://forums.raspberrypi.com/viewtopic.php?t=359302 
I assume we still have a year to make a decision for Raspberry Pi OS. 

But Ubuntu 24.04 will be a problem, it uses Kernel 6.8.

**To everyone: please start new discussions!**

**Reply by mstroh76** [reply]
go to #202

**Comment by Mike941970** [comment]
Hi I am a total noobi working on a battery monitor project using ads1115.c in wiringPi.  I think I have found a bug in the data rate code. According to the data sheet for the device there are 8 data rates, 8SPS, 16SPS, 32SPS, 64SPS, 128SPS, 250SPS, 475SPS, 860 SPS.  Lines 83-89 define 7 bit codes to be inserted into the configuration register. The code for 250SPS is not defined, and the code for 475SPS is defined to have the bit code for 250SPS. the bit code for 860SPS is defined to have the bit code for 475SPS.  On line120 these 7 bit codes are enumerated into dataRates[8].  Since the last member of the array is not enumerated I think its initialized to 0(?) the same as the first member meaning  selecting 860SPS using these defines would select 8SPS. Apparently these defines have been rarely used.
